colcon_core/package_descriptor.py

```diff
@@ -99,26 +99,31 @@ def get_recursive_dependencies(
         """
         queue = self.get_dependencies(categories=direct_categories)
         dependencies = set()
+        depth = 0
         while queue:
-            # pick one dependency from the queue
-            dep = queue.pop()
             # ignore redundant dependencies
-            if dep in dependencies:
-                continue
-            # ignore circular dependencies
-            if dep == self.name:
-                continue
-            # ignore unknown dependencies
-            # explicitly allow multiple packages with the same name
-            descs = [desc for desc in descriptors if desc.name == dep]
-            if not descs:
-                continue
-            # recursing into the same function of the dependency descriptor
-            # queue recursive dependencies
-            for d in descs:
-                queue |= d.get_dependencies(categories=recursive_categories)
-            # add dependency to result set
-            dependencies.add(dep)
+            level_queue = queue - dependencies
+            queue.clear()
+            depth += 1
+            for dep in level_queue:
+                # ignore circular dependencies
+                if dep == self.name:
+                    continue
+                # ignore unknown dependencies
+                # explicitly allow multiple packages with the same name
+                descs = [desc for desc in descriptors if desc.name == dep]
+                if not descs:
+                    continue
+                # recursing into the same function of the dependency descriptor
+                # queue recursive dependencies
+                for d in descs:
+                    queue |= d.get_dependencies(
+                        categories=recursive_categories)
+                # duplicate the descriptor and metadata and add the depth
+                dep = DependencyDescriptor(dep, metadata=dict(dep.metadata))
```

Comment:
Why does the descriptor need to be duplicated?

Reply:
Because some descriptors are created and stored as regular dependencies, so modifying the metadata affects those references to the same instance. Worse yet, those shared references appear in the recursive dependencies of multiple packages at different levels, so we keep overwriting the same 'depth' value.

Reply:
Oh, I see, `get_dependencies()` returns instances it holds on to internally. How about doing a `copy.deepcopy(dep)` here then (to not depend rely on knowledge about the constructor signature)?

Reply:
I actually had started doing exactly that in `get_dependencies()` but decided to do this instead. I'll switch - one moment.